PP-806: Add a new queue attribute named "partition"

suresh-thelkar opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

PP-806
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-806](https://pbspro.atlassian.net/browse/PP-806)**

#### Problem description
* This is one of the tickets of the following EPIC
**[PP-337](https://pbspro.atlassian.net/browse/PP-337)** Multiple schedulers servicing the PBS cluster

This ticket addresses adding new attribute named "partition" to queue object.

#### Cause / Analysis

- Adding new queue attribute named "partition"

#### Solution description

- Adding new queue attribute named "partition"
- "partition" set on routing queues is not allowed
- Changing an execution queue's queue_type to routing when partition is already set on it should not be  allowed.


#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

suresh-thelkar commented:
We have attached PTL test case logs to the JIRA ticket.
https://pbspro.atlassian.net/browse/PP-806

bhroam commented on src/include/pbs_error.h:
```diff
@@ -280,6 +280,8 @@ extern "C" {
 #define PBSE_SCHED_PRIV_EXIST 15215	/* Scheduler sched_priv directory already exists */
 #define PBSE_SCHED_LOG_EXIST 15216	/* Scheduler sched_log directory already exists */
+#define PBSE_ROUTE_QUE_NO_PARTITION  15217 /*Partition can not be assigned to route queue */
+#define PBSE_CANNOT_SET_ROUTE_QUE 15218 /*Can not set queue type to route */
```
Just curious, why did you start here instead of using the gap of numbers starting around 15175?

suresh-thelkar replied:
The last error number that was used before this new error code introduced is PBSE_SCHED_LOG_EXIST(15216). I have continued it from there. 
If you would like us to change these then we have to change all error codes starting form 15214 to the error codes around 15175.   Please provide your thoughts.

bhroam replied:
I'm not sure how this happened, but we have a gap in our error numbers between 15178 to 15200.  Since you added your previous multisched error numbers after 15200, you might as well leave these where they are.

bhroam commented on src/lib/Libattr/attr_fn_str.c:
```diff
@@ -308,3 +308,29 @@ decode_jobname(attribute *patr, char *name, char *rescn, char *val)
+
+int
+set_str_extnd(struct attribute *attr, struct attribute *new, enum batch_op op)
+{
```
I don't understand the need for this function.  The partition attribute is a string.  Why can't you use set_str().  Why do you not want someone to do a += or -=?  While those are seldom used, why block them?  Also your EDD doesn't state this.  Personally I'd rather see you use set_str().

suresh-thelkar replied:
Initially we thought of using set_str() only. Indeed we have done that. But if the user tries to use += or -= unknowingly or as a negative test case then if we use set_str() these operations are allowed(lets say q1 partition=p1 initially and when they do qmgr -c "s q q1 partition += p2" then q1 partition=p1p2 which is not there with any node) and the queue having this unknown partition may not be used for scheduling. By accepting these operations we feel we are telling to users in a way that queue can contain multiple partitions which is not true.

Also observed that this is the behavior with some other attributes like "mail_from". Here also users can use += and  since it does not throw any error it may result in not sending the mail to the original user present in mail_from.

As you said these operations(+=,-=) are seldom used, but we think better not allow them at the first place if these operations are not supported for a particular attribute. Please let us know your thoughts on the same. I will implement it accordingly.

bhroam replied:
While what you say makes sense, this is a problem with our string resources as a whole.  I wouldn't single out partition and do something different.  We should either make an architectural decision to change set_str(), or leave it be.  I don't think partition should be different than other string resources.  This would cause a caveat to be placed in the docs to explain this.

suresh-thelkar replied:
The initial idea of us was that if an attribute does not support operations like += or -= we thought we will disallow them at least when adding new attributes from now onwards.  But as you said since this is a problem with our string resources as a whole, may be it is time to have architectural decision on changing set_str(). 

As suggested I am using set_str() only for now.

bhroam commented on test/tests/interfaces/pbs_queue_partition.py:
```diff
@@ -0,0 +1,209 @@
+
+    def create_queue_partition(
+            self,
+            q_type="e",
```
I'd spell this out: "execution".  The short forms are for lazy admins who don't want to type(like me :).  A test should be more explicit.

suresh-thelkar replied:
Done

bhroam commented on test/tests/interfaces/pbs_queue_partition.py:
```diff
@@ -0,0 +1,209 @@
+                            id="Q1", expect=True)
+
+    def test_set_partition_user_permissions(self):
+        """
```
All tests will always pass when running in API mode.  Is that what you want?

vinodchitrali replied:
YES. 
Here  we need to assert the error code. If the framework returns error code then irrespective of mode, we can  do something like 
self.assertTrue(errorCode, 15007)

But 

```
 self.server.manager(MGR_CMD_UNSET,
                                     QUEUE, 'partition',
                                   id="Q1", runas=TEST_USER) 

```in above code we wont get PBS  error code in any form.   The only way to assert the error code is to assert the exception message.

vinodchitrali replied:
There are some issues in API mode. specially with pbs_manager IFL

bhroam replied:
I'd like to hear @hirenvadalia's thoughts on how to write tests like this.  I dislike the idea that your test will always pass in API mode.  Of course most tests I see that check an exception like this check for the qmgr message and will fail in API mode.

hirenvadalia replied:
AFAIK e.rc is return code PBS command in CLI mode and return code of call to IFL API. So yes, we can do self.assertTrue(e.rc, 15007) which will work in both PTL mode (CLI and API)

suresh-thelkar replied:
Thanks for your reply. We have analyzed the PTL code and following is the observation.

It looks like PTL is using Popen to execute the respective qmgr command in CLI mode. Given below is snapshot of code for the same.
         `try:`
                `       p = Popen(runcmd, bufsize=-1, stdin=stdin, stdout=stdout,
                          stderr=stderr, cwd=cwd, env=env)`
`except Exception, e:`
`....`

Upon Popen execution it is getting the return code from the following statements.
`           (o, e) = p.communicate(input)
               `
` ret['rc'] = p.returncode`

This p.returncode is nothing but the return code of the process that is forked which is qmgr in this case. And it is not returning the PBS error code. Following is an example for the same.

[root@stblr3 tests]# qmgr -c "s q q5 started=t"
qmgr obj=q5 svr=default: Unknown queue
qmgr: Error (15018) returned from server
[root@stblr3 tests]# echo $?
170

[root@stblr3 tests]# qmgr -c 'create queue Q1 started=False,queue_type=route,partition=P1,enabled=False'
qmgr obj=Q1 svr=default: Can not assign a partition to route queue
qmgr: Error (15217) returned from server
[root@stblr3 tests]# echo $?
113
As per our analysis in the above case e.rc is coming as 113 which is indeed not a PBS error code.

Going by this logic we feel whatever you suggested might not work.

suresh-thelkar replied:
Following is further analysis on this.

- The return code from qmgr is manipulated by the underlying shell in case of LINUX operating system i.e. even though we return 15217 from qmgr it is converted to (15217 % 256 = 113) and hence we are seeing the above error code. But in case of Windows it might return the exact error code i.e. 15217
- It would always better for PTL to parse the error message from the object e of the statement
(o, e) = p.communicate(input) and then fill the rc with actual value since we i.e. PTL itself is raising an exception. This way we can have consistency both in LINUX and non LINUX environments.  It is not good to blindly assume that qmgr always returns PBS error code.

hirenvadalia replied:
@suresh-thelkar I think this is issue with qmgr (or in general PBS commands) as their return code is not uniform. If you see my [this](https://github.com/PBSPro/pbspro/pull/324#discussion_r126435140) comment on #324 in that qmgr example it is returning correct return code.

suresh-thelkar replied:
@hirenvadalia 
As mentioned in earlier comments, PTL returns proper error code in case of WINDOWS OS only since the shell on Windows does not manipulate the exit code returned by PBS commands like qmgr. But whereas in case of LINUX it returns (exit_code % 256) as return code.  Please look into the above suggestions made by us in earlier comments.

Also we checked the example you have given. Seems you executed that command on Windows only and hence you are getting the right error code by qmgr.

vinodchitrali replied:
From @suresh-thelkar analysis we can  do some fix in test case. I suggest to have fix in framework side return errorCode in rc.  @kjakkali Please take a look into thread. 

Continuing  @suresh-thelkar analysis. 
rc can have two values   (errorCode) or   (errorCode % 256)
In windows(qmgr CLI)   -  rc  =  (errorCode) -   15217
In Api mode                     -  rc  = (errorCode)  -  15217
In Linux (qmgr CLI)        -  rc  = (errorCode%256) - 113


Now , what will happen when we always assert (errorCode%256)
something like 
      self.assertTrue( (rc%256) == (errorCode%256), "Fail")
------
In windows(qmgr CLI)   -        self.assertTrue( (15217%256) == (15217%256), "Fail")  - self.assertTrue( 113 == 113, "Fail")
In Api mode                     -        self.assertTrue( (15217%256) == (15217%256), "Fail") - self.assertTrue( 113 == 113, "Fail")
In Linux (qmgr CLI)        -        self.assertTrue( (113%256) == (15217%256), "Fail") - self.assertTrue( 113 == 113, "Fail")

So all test cases are protected.  Solves all problems

kjakkali replied:
@vinodchitrali : I agree that framework should take care such things instead of test cases. For PBS commands framework should return PBS error code irrespective of operating system or mode.

bhroam commented on test/tests/interfaces/pbs_queue_partition.py:
```diff
@@ -0,0 +1,209 @@
+        msg3 = "checking the qmgr error message"
+        try:
+            self.create_queue_partition(
+                q_type='r', enable='False', start='False')
```
Same as above.  I'd spell this out to route

suresh-thelkar replied:
Done

bhroam commented on test/tests/interfaces/pbs_queue_partition.py:
```diff
@@ -0,0 +1,209 @@
+    def test_set_partition_without_queue_type(self):
+        """
+        Test to check the set of partition attribute on queue
+         with not queue_type set
```
Is there an extra space before the 'w'?

suresh-thelkar replied:
Done

bhroam commented on test/tests/interfaces/pbs_queue_partition.py:
```diff
@@ -0,0 +1,209 @@
+                            QUEUE, {'partition': "P2"},
+                            id="Q1", expect=True)
+        self.server.manager(MGR_CMD_SET,
+                            QUEUE, {'queue_type': "e"},
```
Same as above, spell out 'execution'

suresh-thelkar replied:
Done

bhroam commented on src/server/req_manager.c:
```diff
@@ -650,6 +651,11 @@ set_queue_type(attribute *pattr, void *pque, int mode)
+			} else {
+				if (((pbs_queue *)pque)->qu_attr[(int)QA_ATR_partition].at_flags & ATR_VFLAG_SET &&
+						(spectype == QTYPE_RoutePush)) {
+					return PBSE_CANNOT_SET_ROUTE_QUE;
```
Won't this get handled by the action function?

suresh-thelkar replied:
The current function is action function only.  It was named as set_queue_type which may be causing the confusion. It could have been renamed to something like action_queue_type.

bhroam commented on src/server/req_manager.c:
```diff
@@ -1686,6 +1692,13 @@ mgr_queue_set(struct batch_request *preq)
 		rc = mgr_set_attr(pque->qu_attr, que_attr_def, QA_ATR_LAST,
 			plist, preq->rq_perm, &bad, (void *)pque,
 			ATR_ACTION_ALTER);
+		if (rc == PBSE_CANNOT_SET_ROUTE_QUE) {
```
Shouldn't this be in set_queue_type() (queue_type's action function)?

suresh-thelkar replied:
The error code PBSE_CANNOT_SET_ROUTE_QUE is being set in set_queue_type() action function only. We thought of logging this in server as well if this error occurs. 

Also when I do not have this block of code here, it was actually appending the name of the attribute to the error message which is queue_type something like 
"Unable to set queue_type to route queue with the partition attribute set **queue_type**"
But the requirement is to exactly get the following message where queue_type should not be appended.
"Unable to set queue_type to route queue with the partition attribute set"

bhroam replied:
I dislike special case code.  I'd rather everything work within the attribute framework.  The mgr_queue_set() function should remain agnostic to what attribute is being set.  It's a generic function.  Instead of having the special case code, I'd rather see the error message be changed to something that works with the existing format.  How about "route queues are incompatible with the partition attribute".  That wouldn't seem weird having queue_type printed at the end.  It would also be kind of weird that all other rejects have the attribute name at the end and this one wouldn't.

suresh-thelkar replied:
As suggested changed the error message to the following.

"Route queues are incompatible with the partition attribute"

Have updated EDD accordingly.

suresh-thelkar commented:
@bhroam  
Thank you very much for quick review. We will implement your comments and get back to you.

suresh-thelkar commented:
@bhroam 
We have addressed all of your comments. Please look into the latest commit for code changes.
Also attached the PTL logs to the ticket after implementing these commits.

Please provide your sign off or further review comments if any.

suresh-thelkar commented:
@bhroam 
Thank you very much for looking into the last commit and reviewing it very quickly.  The approach suggested by you perfectly works fine. We hope you are suggesting this workaround only because the following way of checking the error code as suggested by @hirenvadalia is currently broken in PTL (since PTL framework does not populate e.rc with 15217 in case of LINUX operating systems as explained by me in detail in earlier comments)

`self.assertTrue(e.rc, 15217)`
Note: The above error code is returned when the following command is given to PBS.
`qmgr -c 'create queue Q1 started=False,queue_type=route,partition=P1,enabled=False'`

As per my knowledge I believe that this can be easily fixed in PTL framework if we do the following changes.

- PTL framework  in case of CLI mode, can be changed to parse the error message from the object e of the statement
(o, e) = p.communicate(input) and then fill the rc with actual error code value. Since PTL framework itself is raising the exception it is possible.
- In case of API or IFL mode, since PTL framework uses swigified version of pbs_manager() which is expected to return the right error code as given by PBS, here also PTL framework can raise the exception by populating the right error code in e.rc

If the above things are done in PTL framework itself, test cases code look similar both in CLI and API mode and the test case writers can easily develop their test cases without bothering about the internal framework details. We are actually planning to raise a bug in PTL to handle the same.

suresh-thelkar commented:
@bhroam 
As said earlier we have logged the following PTL framework bugs which will take care of running PTL tests in API mode.
[PP-881](https://pbspro.atlassian.net/browse/PP-881)  PTL framework is not consistent while returning PBS error code through PbsManagerError
[PP-878](https://pbspro.atlassian.net/browse/PP-878)  pbs_swigify is broken

For the time being since all the tests are passing in CLI mode we thought of leaving these test cases as it is.  Once the above bugs get fixed we can revisit these test cases. 

Can you please provide your sign off to proceed further.

bhroam commented:
@suresh-thelkar 
Thanks for filing the PTL bugs.  I'm surprised pbs_swigify didn't work for you.  It worked fine for me.  In any case, we can do a little better until e.rc is correctly returned by PTL.  Don't check for the command output.  You can say self.assertTrue(msg in e.msg) where msg is one of the PBS error messages.  You can also add in the commented out check for e.rc and say it's blocked on PP-881.

suresh-thelkar commented:
@bhroam 
We have fixed all the PTL test scripts as you suggested. Please look into the latest commit and provide your sign off / give further review comments if any.

bhroam commented on test/tests/interfaces/pbs_queue_partition.py:
```diff
@@ -95,41 +95,41 @@ def test_set_unset_partition(self):
         except PbsManagerError as e:
-            self.assertTrue(msg1 in e.msg[0], msg3)
-            self.assertTrue(msg2 in e.msg[1], msg3)
+            # self.assertTrue(e.rc == 15007, msg2)
```
Could you make this assertEquals()? (same throughout the test)

suresh-thelkar replied:
Done

suresh-thelkar commented:
@bhroam 
We have implemented your last round of review comments as well. Can you please provide your sign off/further review comments if any.

suresh-thelkar commented:
@PBSPro/pbspro-maintainers 
Have got all necessary sign offs. Also squashed all the commits and rebased.  Please it take it forward.
